kernel/src/fs/notify/inotify.rs

```diff
@@ -0,0 +1,607 @@
+// SPDX-License-Identifier: MPL-2.0
+
+use alloc::{
+    collections::VecDeque,
+    string::String,
+    sync::{Arc, Weak},
+};
+use core::{
+    any::Any,
+    sync::atomic::{AtomicBool, AtomicU32, AtomicU64, Ordering},
+};
+
+use bitflags::bitflags;
+use hashbrown::HashMap;
+use ostd::{mm::VmWriter, sync::Mutex};
+
+use crate::{
+    current_userspace,
+    events::IoEvents,
+    fs::{
+        file_handle::FileLike,
+        notify::{FsnotifyEvents, FsnotifySubscriber},
+        path::Path,
+        pseudofs::anon_inodefs_shared_inode,
+        utils::{AccessMode, Inode, IoctlCmd, StatusFlags},
+    },
+    prelude::*,
+    process::signal::{PollHandle, Pollable, Pollee},
+    return_errno_with_message,
+};
+
+#[derive(Clone)]
+struct SubscriberEntry {
+    inode: Weak<dyn Inode>,
+    subscriber: Weak<dyn FsnotifySubscriber>,
+}
+
+/// A file-like object that provides inotify functionality.
+///
+/// InotifyFile accepts events from multiple inotify subscribers (watches) on different inodes.
+/// Users should read events from this file to receive notifications about filesystem changes.
+pub struct InotifyFile {
+    // Lock to serialize watch updates and removals.
+    watch_lock: Mutex<()>,
+    // Next watch descriptor to allocate.
+    next_wd: AtomicU32,
+    // A map from watch descriptor to subscriber entry.
+    watch_map: RwLock<HashMap<u32, SubscriberEntry>>,
+    // Whether the file is opened in non-blocking mode.
+    is_nonblocking: AtomicBool,
+    // Bounded queue of inotify events.
+    event_queue: RwLock<VecDeque<InotifyEvent>>,
+    // Maximum capacity of the event queue.
+    queue_capacity: usize,
+    // A weak reference to this inotify file.
+    this: Weak<InotifyFile>,
+    // A pollable object for this inotify file.
+    pollee: Pollee,
+}
+
+impl Drop for InotifyFile {
+    /// Cleans up all subscribers when the inotify file is dropped.
+    /// This will remove all subscribers from their inodes.
+    fn drop(&mut self) {
+        let watch_map = self.watch_map.write();
+        for (_, entry) in watch_map.iter() {
+            // The weak refs may have already been dropped by concurrent activity.
+            // If so, skip them rather than unwrapping.
+            if let (Some(inode), Some(subscriber)) =
+                (entry.inode.upgrade(), entry.subscriber.upgrade())
```

Comment:
Reduce the indentation level by using `let-else`.

```rust
let (Some(inode), Some(subscriber)) = (entry.inode.upgrade(), entry.subscriber.upgrade()) else {
    continue;
};

inode.fsnotify_publisher().remove_subscriber(&subscriber);
inode.fs().fsnotify_info().remove_subscriber();
```

In general, we should alway prefer the coding patterns that handle the error logic--rather than the main logic--in `if` branches.